### ISSUE TITLE
Detect that adapter is not connected

### ISSUE THREAD

**Issue by HCanber** [tcdl/msb#56]
We have no way of detecting if an adapter is connected to its queue, e.g. RabbitMQ or Redis. The callbacks we provide are not called with an error, as the publications are queued, waiting for the source to come back up. When the queue comes online, the queued publications are sent.

In most scenarios, when the application starts we will immediately connect to the queue so the behavior described above is perfectly fine if the queue then suddenly dies.
But in scenarios when we did not connect to the queue when the app initialized (or within 30 seconds), we want to get a notification about this. It could be because the queue is down, but it could also mean we have invalid config and is trying to connect to a server that do not exist.

And, to be able to produce status reports about the application, we would also like to tell if we're currently connected to the queue.
#### Suggestion

As both [redis npm package](https://www.npmjs.com/package/redis#ready) and [amqp-coffe](https://www.npmjs.com/package/amqp-coffee) expose events, we could make the channel objects into EventEmitters (I know subscriber channel already is), and whenever the underlying adapter publishes it's connected or disconnected we let the channel publish `ready` and `closed` events.

Based on those we can provide the status reports needed, and we can send an alarm if we do not receive a `ready` event within 30 s after application start.

Thoughts?

**Comment by jupiter** [comment]
Yep, it is rather annoying not knowing what's happening with the connection.

However, a connection is not necessarily attempted on the creation of a channelManager, but rather when the first channel is created. Both Redis and AMQP libraries handles automatic reconnection, as well as connection to alternative host based on availability. This means that the 'ready' and 'close'/'end' events will not be easily actionable, other than for logging.

I therefore propose that we add a 'connectionTimeout' event, which will fire only after a configurable number of ms if multiple connection attempts had failed. It would also fire if a reconnection attempt had failed.

**Comment by HCanber** [comment]
It's for logging and dashboards that I want to know if the adapter is connected or not. I don't want to take any other action than firing off an alarm. 

As we use start `channelMonitorAgent` on application startup a connection attempt will be made immediately.

Although a `connectionTimeout` event might be useful it will not answer what state the adapter is in.

**Comment by jupiter** [comment]
Some questions:
- When you get the status change, where will you be reporting it?
- Would you want to fire 'an alarm' on any reconnection attempt, incl. a change of broker host?
- Would we want to use the status to provide some form of flow control or error feedback to publishing operations?

**Comment by HCanber** [comment]
> When you get the status change, where will you be reporting it?

Probably just log it, and have something looking it, or just put it in a graph in Kibana.
We have also been discussing querying the service for some kind of health/status report, which would report `connectedToBus: true/false`

> Would you want to fire 'an alarm' on any reconnection attempt, incl. a change of broker host?

Maybe not 'an alarm' as in something-has-gone-wrong, but I'm certainly interested in logging that we've lost the connection. The connection attempt per se, is not that interesting, but the _end/disconnected_ and _ready/connected_ events that precede and comes after the connection attempt are.

> Would we want to use the status to provide some form of flow control or error feedback to publishing operations?

Maybe. It might be used in a circuit breaker to force it to open and let publishing operations to fail instead of just hanging. I guess the answer is: it depends `:)`
